barretenberg/cpp/src/barretenberg/honk/proof_system/logderivative_library.hpp

```diff
@@ -39,26 +40,33 @@ void compute_logderivative_inverse(Polynomials& polynomials, auto& relation_para
     constexpr size_t WRITE_TERMS = Relation::WRITE_TERMS;
 
     auto& inverse_polynomial = Relation::template get_inverse_polynomial(polynomials);
-    for (size_t i = 0; i < circuit_size; ++i) {
-        // TODO(https://github.com/AztecProtocol/barretenberg/issues/940): avoid get_row if possible.
-        auto row = polynomials.get_row(i);
-        bool has_inverse = Relation::operation_exists_at_row(row);
-        if (!has_inverse) {
-            continue;
-        }
-        FF denominator = 1;
-        bb::constexpr_for<0, READ_TERMS, 1>([&]<size_t read_index> {
-            auto denominator_term =
-                Relation::template compute_read_term<Accumulator, read_index>(row, relation_parameters);
-            denominator *= denominator_term;
-        });
-        bb::constexpr_for<0, WRITE_TERMS, 1>([&]<size_t write_index> {
-            auto denominator_term =
-                Relation::template compute_write_term<Accumulator, write_index>(row, relation_parameters);
-            denominator *= denominator_term;
-        });
-        inverse_polynomial.at(i) = denominator;
-    };
+    const size_t min_iterations_per_thread = 128;
+    size_t num_threads = bb::calculate_num_threads_pow2(circuit_size, min_iterations_per_thread);
+    const size_t rows_per_thread = circuit_size / num_threads;
+    parallel_for(num_threads, [&](size_t thread_idx) {
```

Comment:
Just a comment that the caller of this function might be multithreading (e.g., per relation). Maybe BB doesn't but we do: https://github.com/AztecProtocol/aztec-packages/blob/master/barretenberg/cpp/src/barretenberg/vm2/constraining/prover.cpp#L66

IIRC, this will make it crash since `parallel_for` cannot be called recursively.

If you want I can run some VM benchmarking before and after this PR to see what happens if we disable our parallelism over relations. For a long enough trace the approach in this PR would probably be better. (ping me if I forget)

PS: Long term, the AVM might diverge from using this function, because our witgen can output metadata with the exact set of rows where every lookup/permutation is active.

Reply:
This seems to be strictly worse for the AVM. Probably because we have enough columns to nicely saturate 1 thread per column (while achieving good cache locality per thread).

```
MASTER

(avm, 16 cores) prove/execute_log_derivative_inverse_round_ms: 1224
(avm, 32 cores)prove/execute_log_derivative_inverse_round_ms: 637

THIS PR

(avm, 16 cores) prove/execute_log_derivative_inverse_round_ms: 1663
(avm, 32 cores) prove/execute_log_derivative_inverse_round_ms: 1338```

Reply:
Hmm, not that I know (cc: @ludamad)

Reply:
we could expose the 'nested' variable from parallel_for_mutex_pool